File: modules/google-earth-engine/docker/sepal-ee/sepal/ee/rx/task.py
import ee
from ee.batch import Task
from rx import empty, of
from rx.operators import distinct_until_changed, finally_action, flat_map, take_while
from sepal.rx import throw
from sepal.task.rx.observables import progress

from .observables import execute, interval

_MONITORING_FREQUENCY = 5


def execute_task(credentials, task):
    def start():
        task.start()

    def load_status():
        return task.status()

    def extract_state(status):
        state = status['state']
        if state == 'FAILED':
            return throw(ee.EEException(status.get('error_message')))
        else:
            return of(state)

    def to_progress(state):
        if state == 'PENDING':
            return progress(
                default_message='Submitting export task to Google Earth Engine...',
                message_key='tasks.ee.export.pending'
            )
        elif state == 'READY':
            return progress(
                default_message='Waiting for Google Earth Engine to start export...',
                message_key='tasks.ee.export.ready'
            )
        elif state == 'RUNNING':
            return progress(
                default_message='Google Earth Engine is exporting image...',
                message_key='tasks.ee.export.running'
            )
        else:
            return empty()

    def monitor():
        def is_running(state):
            return state in [Task.State.UNSUBMITTED, Task.State.READY, Task.State.RUNNING]

        return interval(credentials, _MONITORING_FREQUENCY).pipe(
            flat_map(
                lambda _: execute(
                    credentials,
                    action=load_status,
                    description='monitor task ' + str(task),
                    retries=3
                )
            ),
            flat_map(extract_state),
            distinct_until_changed(),
            take_while(is_running, inclusive=True),
            flat_map(to_progress)
        )

    return execute(
        credentials,
        start,
        description='start task ' + str(task),
        retries=3
    ).pipe(
        flat_map(lambda _: monitor()),
        finally_action(lambda: task.cancel())
    )

File: modules/task/docker/src/ee/task.js
const ee = require('ee')
const {interval, of, throwError} = require('rxjs')
const {distinctUntilChanged, exhaustMap, first, finalize, map, switchMap, mapTo, takeWhile, tap} = require('rxjs/operators')
const log = require('sepal/log').getLogger('ee')

const MONITORING_FREQUENCY = 10000
const {UNSUBMITTED, READY, RUNNING, FAILED} = ee.data.ExportState

const executeTask$ = (task, description) => {
    const monitor$ = task =>
        interval(MONITORING_FREQUENCY).pipe(
            exhaustMap(() => status$(task)),
            switchMap(({state, error_message: error}) => error || state === FAILED
                ? throwError(new Error(error))
                : of(state)
            ),
            distinctUntilChanged(),
            takeWhile(state => isRunning(state)),
            map(toProgress)
        )

    const toProgress = state => {
        switch (state) {
            case 'UNSUBMITTED':
                return {
                    state,
                    defaultMessage: 'Submitting export task to Google Earth Engine',
                    messageKey: 'tasks.ee.export.pending'
                }
            case 'READY':
                return {
                    state,
                    defaultMessage: 'Waiting for Google Earth Engine to start export',
                    messageKey: 'tasks.ee.export.ready'
                }
            case 'RUNNING':
                return {
                    state,
                    defaultMessage: 'Google Earth Engine is exporting',
                    messageKey: 'tasks.ee.export.running'
                }
            default:
                throw Error(`Unknown state: ${state}`)
        }
    }

    const cleanup = task => {
        status$(task).pipe(
            map(({state}) => isRunning(state)),
            switchMap(running => running
                ? cancel$(task).pipe(
                    mapTo(true)
                )
                : of(false)
            ),
            first()
        ).subscribe({
            next: wasRunning => log.info(`EE task ${task.id}: ${wasRunning ? 'Cancelled' : 'Completed'} (${description})`),
            error: error => log.error('Failed to cancel EE task', error)
        })
    }

    const start$ = task =>
        ee.$({
            operation: `start task (${description})`,
            ee: (resolve, reject) =>
                ee.data.startProcessing(null, task.config_, (result, error) => {
                    if (error) {
                        reject(error)
                    } else {
                        task.id = result.taskId // [TODO] Solve this without mutation
                        resolve()
                    }
                })
        }).pipe(
            tap(() => log.info(`EE task ${task.id}: Started (${description})`))
        )

    const status$ = task =>
        ee.$({
            operation: `get task status (${description})`,
            ee: (resolve, reject) =>
                ee.data.getTaskStatus(task.id,
                    (status, error) => error ? reject(error) : resolve(status)
                )
        }).pipe(
            map(([status]) => status)
        )

    const cancel$ = task =>
        ee.$({
            operation: `cancel task (${description})`,
            ee: (resolve, reject) =>
                ee.data.cancelTask(task.id,
                    (_canceled, error) => error ? reject(error) : resolve()
                )
        })

    const isRunning = state => [UNSUBMITTED, READY, RUNNING].includes(state)

    return start$(task).pipe(
        switchMap(() => monitor$(task)),
        finalize(() => task.id && cleanup(task))
    )
}

module.exports = {executeTask$}
